From 0b8c48688fdcbabe957a08b3c80402ba0fe4e626 Mon Sep 17 00:00:00 2001
From: "Alan T. DeKok" <aland@freeradius.org>
Date: Sun, 2 Jul 2017 20:54:36 -0400
Subject: [PATCH] FR-GV-205 - check for "too long" options, too

---
 src/lib/dhcp.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/lib/dhcp.c b/src/lib/dhcp.c
index 0790a0d32..72c34088b 100644
--- a/src/lib/dhcp.c
+++ b/src/lib/dhcp.c
@@ -600,6 +600,12 @@ ssize_t fr_dhcp_decode_options(uint8_t *data, size_t len, VALUE_PAIR **head)
 			continue;
 		}
 
+		/*
+		 *	End of this attribute past the end of the
+		 *	packet: ignore it.
+		 */
+		if (next > (data + len)) break;
+
 		da = dict_attrbyvalue(DHCP2ATTR(p[0]));
 		if (!da) {
 			fr_strerror_printf("Attribute not in our dictionary: %u",
-- 
2.13.2

