From bc084c4df2e7b2091e94811e8930b1bb2e4228cf Mon Sep 17 00:00:00 2001
From: "Alan T. DeKok" <aland@freeradius.org>
Date: Mon, 3 Jul 2017 11:35:02 -0400
Subject: [PATCH] FR-GV-206 - decode option 60 (string) not 63 (octets)

---
 src/lib/dhcp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/lib/dhcp.c b/src/lib/dhcp.c
index 72c34088b..b8d78ffef 100644
--- a/src/lib/dhcp.c
+++ b/src/lib/dhcp.c
@@ -821,7 +821,7 @@ int fr_dhcp_decode(RADIUS_PACKET *packet)
 			/*
 			 *	Vendor is "MSFT 98"
 			 */
-			vp = pairfind(head, DHCP2ATTR(63));
+			vp = pairfind(head, DHCP2ATTR(60));
 			if (vp && (strcmp(vp->vp_strvalue, "MSFT 98") == 0)) {
 				vp = pairfind(head, DHCP2ATTR(262));
 
-- 
2.13.2

