From 7c59bf2d6afce15ee66c54c6382c5afdfbbfd344 Mon Sep 17 00:00:00 2001
From: "Alan T. DeKok" <aland@freeradius.org>
Date: Sat, 1 Jul 2017 08:49:24 -0400
Subject: [PATCH] FR-GV-204 - free VP if decoding options fails, so we don't
 leak memory

---
 src/lib/dhcp.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/lib/dhcp.c b/src/lib/dhcp.c
index 040648017..0790a0d32 100644
--- a/src/lib/dhcp.c
+++ b/src/lib/dhcp.c
@@ -793,6 +793,7 @@ int fr_dhcp_decode(RADIUS_PACKET *packet)
 	 */
 	if (fr_dhcp_decode_options(packet->data + 240, packet->data_len - 240,
 				   tail) < 0) {
+		pairfree(&head);
 		return -1;
 	}
 
@@ -1452,6 +1453,7 @@ int fr_dhcp_encode(RADIUS_PACKET *packet)
 			 *	limitations: sizeof(vp->vp_octets) < 255
 			 */
 			if (length > 255) {
+				pairfree(&vp);
 				fr_strerror_printf("WARNING Ignoring too long attribute %s!", vp->name);
 				break;
 			}
-- 
2.13.2

